File: problems/0049-group-anagrams/solution.cpp
#include <array>
#include <functional>
#include <string>
#include <vector>


namespace std {
    // AI-generated hasher for formula
    template<>
    struct hash<std::array<unsigned char, 26>> {
        size_t operator()(const std::array<unsigned char, 26>& arr) const {
            size_t h = 0;
            for (auto val : arr) {
                h ^= std::hash<unsigned char>{}(val)+0x9e3779b9 + (h << 6) + (h >> 2);
            }
            return h;
        }
    };
}

class Solution {
public:

    size_t hash_string(const std::string& s) {
        std::array<unsigned char, 26> count = { 0 };
        for (char c : s) {
            count[c - 'a']++;
        }
        std::hash<std::array<unsigned char, 26>> hasher;
        return hasher(count);
    }

    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string>& strs) {
        std::unordered_map<size_t, std::vector<std::string>> map_groups;
        for (auto& s : strs) {
            size_t hash = Solution::hash_string(s);
            map_groups[hash].push_back(std::move(s));
        }
        std::vector<std::vector<std::string>> groups;
        for (auto& [hash, anagrams] : map_groups) {
            groups.push_back(std::move(anagrams));
        }
        return groups;
    }
};

int main([[maybe_unused]] int argc, [[maybe_unused]] char* argv[]) {
    return 0;
}

File: problems/0049-group-anagrams/solution.rs
#![allow(unused)]

use std::collections::HashMap;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
pub struct Solution;

impl Solution {
    pub fn hash_string(s: &str) -> u64 {
        let mut count = [0u8; 26];
        for c in s.bytes() {
            count[(c - b'a') as usize] += 1;
        }
        let mut hasher = DefaultHasher::new();
        count.hash(&mut hasher);
        hasher.finish()
    }

    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut groups: Vec<Vec<String>> = vec![];
        let mut map_groups: HashMap<u64, Vec<String>> = HashMap::new();

        for string in strs {
            let hash = Solution::hash_string(&string);
            map_groups.entry(hash).or_default().push(string);
        }

        map_groups.into_values().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_solution() {
        let result = Solution::group_anagrams(vec!["a".to_string()]);
        assert_eq!(result, vec![vec!["a".to_string()]]);
    }
}
